src/libos/src/process/syscalls.rs

```diff
@@ -142,20 +144,43 @@ pub fn do_futex(
         Ok(val as usize)
     };
 
-    let get_futex_timeout = |timeout| -> Result<Option<FutexTimeout>> {
-        let timeout = timeout as *const timespec_t;
-        if timeout.is_null() {
-            return Ok(None);
-        }
-        let ts = timespec_t::from_raw_ptr(timeout)?;
-        ts.validate()?;
-        // TODO: use a secure clock to transfer the real time to monotonic time
-        let clock_id = if futex_flags.contains(FutexFlags::FUTEX_CLOCK_REALTIME) {
-            ClockID::CLOCK_REALTIME
+    let get_futex_timeout = |timeout| -> Result<Option<Duration>> {
+        // Sanity checks
+        let mut timeout: Duration = {
+            let timeout = timeout as *const timespec_t;
+            if timeout.is_null() {
+                return Ok(None);
+            }
+            let ts = timespec_t::from_raw_ptr(timeout)?;
+            ts.validate()?;
+            ts.as_duration()
+        };
+
+        // Only FUTEX_WAIT takes the timeout input as relative
+        let is_absolute = if futex_op != FutexOp::FUTEX_WAIT {
+            true
         } else {
-            ClockID::CLOCK_MONOTONIC
+            false
         };
-        Ok(Some(FutexTimeout::new(clock_id, ts)))
+        if is_absolute {
+            // TODO: use a secure clock to transfer the real time to monotonic time
+            let clock_id = if futex_flags.contains(FutexFlags::FUTEX_CLOCK_REALTIME) {
+                ClockID::CLOCK_REALTIME
+            } else {
+                ClockID::CLOCK_MONOTONIC
+            };
+            let now = crate::time::do_clock_gettime(clock_id)
+                .unwrap()
+                .as_duration();
+            timeout = timeout
+                .checked_sub(now)
+                .ok_or_else(|| errno!(EINVAL, "timeout is invalid"))?;
```

Comment:
I think we should return ETIMEDOUT here because `timeout` is already validated here. If the `checked_sub` fails, this means the timeout is expired.